### PR TITLE
PHPC-1716 Allow configuring server API version in manager

### PR DESCRIPTION
**Pull request by alcaeus**
PHPC-1716

This PR is built on the current WIP state of versioned API support in the C driver to allow designing the PHP driver interface and verify that it passes all spec tests. Note that the changes to submodules (including the libmongoc version) will be re-done once versioned API support lands in the main libmongoc development branch.

### CODE REVIEW

File: src/MongoDB/ServerApi.c
Comment (jmikola) [inline src/MongoDB/ServerApi.c:42]: Is this ever reached? If not, would it be better to throw an exception and return `false`? AFAICT, this is only called from the constructor and `php_phongo_serverapi_init_from_hash`, both of which should use freshly initialized objects.

File: src/MongoDB/ServerApi.c
Comment (jmikola) [inline src/MongoDB/ServerApi.c:112]: Does PHP core indent these blocks? I realize that might not play nice with `clang-format`.

File: src/MongoDB/ServerApi.c
Comment (jmikola) [reply src/MongoDB/ServerApi.c:204]: Looking now at what you did in the `__construct` method, perhaps this is OK and you were just missing the error handling override?

File: src/MongoDB/ServerApi.c
Comment (jmikola) [inline src/MongoDB/ServerApi.c:269]: Is the semicolon after `return` purposely omitted? I know the macro adds its own, but for our readability perhaps it makes sense to write this as:

```c
ZEND_PARSE_PARAMETERS_END_EX(
    zend_restore_error_handling(&error_handling);
    return;
);
```

File: tests/utils/tools.php
Comment (jmikola) [inline tests/utils/tools.php:97]: Top of the file has `use MongoDB\Driver\Manager;` so this can just be Manager.

File: tests/utils/tools.php
Comment (jmikola) [inline tests/utils/tools.php:94]: Consider a `use` statement at the top, as we do for other classes.

File: .evergreen/config.yml
Comment (jmikola) [inline .evergreen/config.yml:595]: Is this related to https://github.com/mongodb/mongo-php-driver/pull/1206?

File: tests/bson-corpus/array-decodeError-001.phpt
Comment (alcaeus) [reply tests/bson-corpus/array-decodeError-001.phpt:10]: Yes, I replaced all instances of `tools.php` to require the `.inc` file for consistency. I can revert this change if you think that's unwise.

File: .gitmodules
Comment (alcaeus) [reply .gitmodules:3]: I had not managed to create the PR before. https://github.com/mongodb/mongo-c-driver/pull/765 contains the changes necessary, most notably fixing a wrong function name and adding a getter for the `version` field in `mongoc_version_api_t`.

File: .evergreen/config.yml
Comment (alcaeus) [reply .evergreen/config.yml:595]: Yes, this is how I discovered this. I'll rebase this PR once the original has merged, which will also take care of the remaining test failures.

File: src/MongoDB/ServerApi.c
Comment (alcaeus) [reply src/MongoDB/ServerApi.c:2]: Done and created PHPC-1792 to correct this across all files.

File: src/MongoDB/ServerApi.c
Comment (alcaeus) [reply src/MongoDB/ServerApi.c:112]: Unfortunately, clang-format does not allow us to indent these blocks and changes them right back.

File: src/MongoDB/ServerApi.c
Comment (alcaeus) [reply src/MongoDB/ServerApi.c:269]: I omitted it on purpose since the macro appends its own, but after seeing what clang-format does when we do have a semicolon (the resulting line is `return;);` I decided to extract all this functionality into corresponding `PHONGO_PARSE_PARAMETERS` macros which take care of calling `zend_replace_error_handling` appropriately.

File: src/MongoDB/ServerApi.c
Comment (alcaeus) [reply src/MongoDB/ServerApi.c:42]: Replaced this with an exception instructing users to file a bug report as it shouldn't happen.

File: tests/utils/tools.php
Comment (alcaeus) [reply tests/utils/tools.php:97]: You are correct...the joy of not having PHP code assist in my IDE.

File: tests/utils/tools.php
Comment (alcaeus) [reply tests/utils/tools.php:94]: Fixed

File: src/MongoDB/ServerApi.c
Comment (jmikola) [reply src/MongoDB/ServerApi.c:43]: Should we standardize on American English spelling ("initialized") here?

File: phongo_compat.h
Comment (jmikola) [inline phongo_compat.h:187]: Should this be indented consistently with the lines above since it falls within `do {`?

File: phongo_compat.h
Comment (jmikola) [inline phongo_compat.h:190]: Same question about indentation here.

File: tests/bson-corpus/array-decodeError-001.phpt
Comment (jmikola) [reply tests/bson-corpus/array-decodeError-001.phpt:10]: It's not necessary but I can see how it'd make maintenance easier by not having to think about which file to include. No objections to leaving as-is.

**Review by jmikola — APPROVED** [review]
Some questions on code formatting for macros, but LGTM otherwise.

File: phongo_compat.h
Comment (alcaeus) [reply phongo_compat.h:187]: You're right, totally missed that. Thanks!

Comment (alcaeus) [reply phongo_compat.h:187]: Ah...I was wondering about this: it seems like clang-format immediately wants to undo my changes and restore this broken formatting, so I'm afraid we're stuck with this for the time being :(

File: src/MongoDB/ServerApi.c
Comment (alcaeus) [reply src/MongoDB/ServerApi.c:43]: I forgot we use Simplified English as opposed to Traditional English. I'll update it.

**Comment by alcaeus** [comment]
> Beyond that, take a look at the CI failures. AppVeyor reports a compilation failure in some macro.

This was due to me using `ZEND_STRL` in `zend_hash_str_add`, which is implemented as a macro in PHP < 7.3. I've removed this usage for the time being. According to @cmb69, `ZEND_STRL` should not be used; this will most likely also affect other usages of `ZEND_STRL` in similar instances.

> The Evergreen failure appears to be related to what you fixed in #1206.

Yes, those are fixed as well now.

**Comment by alcaeus** [comment]
Note: holding off on merge until https://github.com/mongodb/mongo-c-driver/pull/765 is merged.

**Comment by jmikola** [comment]
> This was due to me using ZEND_STRL in zend_hash_str_add, which is implemented as a macro in PHP < 7.3. I've removed this usage for the time being. According to @cmb69, ZEND_STRL should not be used; this will most likely also affect other usages of ZEND_STRL in similar instances.

I'm familiar with that. I assume it's because `ZEND_STRL` may not be expanded before the preprocessor checks arguments for the outer macro (`zend_hash_str_add` in this case).

Our previous issues with `ZEND_STRL` had to do with it only being suitable for literals (see: https://github.com/mongodb/mongo-php-driver/search?q=ZEND_STRL&type=commits). That seems unrelated to this.

I don't think we need to stop using `ZEND_STRL` altogether, though. The build failures when we do something wrong should be obvious :)

**Pull request merged**
